// mattermost-redux/lib/action_types/index.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedChannelTypes = exports.ScheduledPostTypes = exports.ChannelBookmarkTypes = exports.PlaybookType = exports.DraftTypes = exports.HostedCustomerTypes = exports.ThreadTypes = exports.AppsTypes = exports.CloudTypes = exports.ChannelCategoryTypes = exports.PluginTypes = exports.BotTypes = exports.GroupTypes = exports.SchemeTypes = exports.RoleTypes = exports.SearchTypes = exports.LimitsTypes = exports.JobTypes = exports.AdminTypes = exports.EmojiTypes = exports.IntegrationTypes = exports.PreferenceTypes = exports.FileTypes = exports.PostTypes = exports.ChannelTypes = exports.TeamTypes = exports.UserTypes = exports.GeneralTypes = exports.ErrorTypes = void 0;
const admin_1 = __importDefault(require("./admin"));
exports.AdminTypes = admin_1.default;
const apps_1 = __importDefault(require("./apps"));
exports.AppsTypes = apps_1.default;
const bots_1 = __importDefault(require("./bots"));
exports.BotTypes = bots_1.default;
const channel_bookmarks_1 = __importDefault(require("./channel_bookmarks"));
exports.ChannelBookmarkTypes = channel_bookmarks_1.default;
const channel_categories_1 = __importDefault(require("./channel_categories"));
exports.ChannelCategoryTypes = channel_categories_1.default;
const channels_1 = __importDefault(require("./channels"));
exports.ChannelTypes = channels_1.default;
const cloud_1 = __importDefault(require("./cloud"));
exports.CloudTypes = cloud_1.default;
const drafts_1 = __importDefault(require("./drafts"));
exports.DraftTypes = drafts_1.default;
const emojis_1 = __importDefault(require("./emojis"));
exports.EmojiTypes = emojis_1.default;
const errors_1 = __importDefault(require("./errors"));
exports.ErrorTypes = errors_1.default;
const files_1 = __importDefault(require("./files"));
exports.FileTypes = files_1.default;
const general_1 = __importDefault(require("./general"));
exports.GeneralTypes = general_1.default;
const groups_1 = __importDefault(require("./groups"));
exports.GroupTypes = groups_1.default;
const hosted_customer_1 = __importDefault(require("./hosted_customer"));
exports.HostedCustomerTypes = hosted_customer_1.default;
const integrations_1 = __importDefault(require("./integrations"));
exports.IntegrationTypes = integrations_1.default;
const jobs_1 = __importDefault(require("./jobs"));
exports.JobTypes = jobs_1.default;
const limits_1 = __importDefault(require("./limits"));
exports.LimitsTypes = limits_1.default;
const playbooks_1 = __importDefault(require("./playbooks"));
exports.PlaybookType = playbooks_1.default;
const plugins_1 = __importDefault(require("./plugins"));
exports.PluginTypes = plugins_1.default;
const posts_1 = __importDefault(require("./posts"));
exports.PostTypes = posts_1.default;
const preferences_1 = __importDefault(require("./preferences"));
exports.PreferenceTypes = preferences_1.default;
const roles_1 = __importDefault(require("./roles"));
exports.RoleTypes = roles_1.default;
const schemes_1 = __importDefault(require("./schemes"));
exports.SchemeTypes = schemes_1.default;
const scheudled_posts_1 = __importDefault(require("./scheudled_posts"));
exports.ScheduledPostTypes = scheudled_posts_1.default;
const search_1 = __importDefault(require("./search"));
exports.SearchTypes = search_1.default;
const shared_channels_1 = __importDefault(require("./shared_channels"));
exports.SharedChannelTypes = shared_channels_1.default;
const teams_1 = __importDefault(require("./teams"));
exports.TeamTypes = teams_1.default;
const threads_1 = __importDefault(require("./threads"));
exports.ThreadTypes = threads_1.default;
const users_1 = __importDefault(require("./users"));
exports.UserTypes = users_1.default;
