// mattermost-redux/lib/action_types/integrations.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const key_mirror_1 = __importDefault(require("mattermost-redux/utils/key_mirror"));
exports.default = (0, key_mirror_1.default)({
    RECEIVED_INCOMING_HOOK: null,
    RECEIVED_INCOMING_HOOKS: null,
    RECEIVED_INCOMING_HOOKS_TOTAL_COUNT: null,
    DELETED_INCOMING_HOOK: null,
    RECEIVED_OUTGOING_HOOK: null,
    RECEIVED_OUTGOING_HOOKS: null,
    DELETED_OUTGOING_HOOK: null,
    RECEIVED_CUSTOM_TEAM_COMMANDS: null,
    RECEIVED_COMMAND: null,
    RECEIVED_COMMANDS: null,
    RECEIVED_COMMAND_TOKEN: null,
    DELETED_COMMAND: null,
    RECEIVED_OAUTH_APP: null,
    RECEIVED_OAUTH_APPS: null,
    DELETED_OAUTH_APP: null,
    RECEIVED_APPS_OAUTH_APP_IDS: null,
    RECEIVED_APPS_BOT_IDS: null,
    RECEIVED_OUTGOING_OAUTH_CONNECTIONS: null,
    RECEIVED_OUTGOING_OAUTH_CONNECTION: null,
    DELETED_OUTGOING_OAUTH_CONNECTION: null,
    RECEIVED_DIALOG_TRIGGER_ID: null,
    RECEIVED_DIALOG_ARGUMENTS: null,
    RECEIVED_DIALOG: null,
});
