// mattermost-redux/lib/constants/permissions_sysconsole.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceToSysConsolePermissionsTable = exports.RESOURCE_KEYS = void 0;
const permissions_1 = __importDefault(require("./permissions"));
exports.RESOURCE_KEYS = {
    ABOUT: {
        EDITION_AND_LICENSE: 'about.edition_and_license',
    },
    REPORTING: {
        SITE_STATISTICS: 'reporting.site_statistics',
        TEAM_STATISTICS: 'reporting.team_statistics',
        SERVER_LOGS: 'reporting.server_logs',
    },
    USER_MANAGEMENT: {
        USERS: 'user_management.users',
        GROUPS: 'user_management.groups',
        TEAMS: 'user_management.teams',
        CHANNELS: 'user_management.channels',
        PERMISSIONS: 'user_management.permissions',
        SYSTEM_ROLES: 'user_management.system_roles',
    },
    SYSTEM_ATTRIBUTES: {
        USER_ATTRIBUTES: 'system_attributes.user_attributes',
        ATTRIBUTE_BASED_ACCESS_CONTROL: 'system_attributes.attribute_based_access_control',
    },
    AUTHENTICATION: {
        SIGNUP: 'authentication.signup',
        EMAIL: 'authentication.email',
        PASSWORD: 'authentication.password',
        MFA: 'authentication.mfa',
        LDAP: 'authentication.ldap',
        SAML: 'authentication.saml',
        OPENID: 'authentication.openid',
        GUEST_ACCESS: 'authentication.guest_access',
    },
    INTEGRATIONS: {
        INTEGRATION_MANAGEMENT: 'integrations.integration_management',
        BOT_ACCOUNTS: 'integrations.bot_accounts',
        GIF: 'integrations.gif',
        CORS: 'integrations.cors',
    },
    COMPLIANCE: {
        DATA_RETENTION_POLICY: 'compliance.data_retention_policy',
        COMPLIANCE_EXPORT: 'compliance.compliance_export',
        COMPLIANCE_MONITORING: 'compliance.compliance_monitoring',
        CUSTOM_TERMS_OF_SERVICE: 'compliance.custom_terms_of_service',
    },
    PRODUCTS: {
        BOARDS: 'boards',
    },
    SITE: {
        CUSTOMIZATION: 'site.customization',
        LOCALIZATION: 'site.localization',
        USERS_AND_TEAMS: 'site.users_and_teams',
        NOTIFICATIONS: 'site.notifications',
        ANNOUNCEMENT_BANNER: 'site.announcement_banner',
        EMOJI: 'site.emoji',
        POSTS: 'site.posts',
        FILE_SHARING_AND_DOWNLOADS: 'site.file_sharing_and_downloads',
        PUBLIC_LINKS: 'site.public_links',
        NOTICES: 'site.notices',
        IP_FILTERING: 'site.ip_filters',
    },
    EXPERIMENTAL: {
        FEATURES: 'experimental.features',
        FEATURE_FLAGS: 'experimental.feature_flags',
        BLEVE: 'experimental.bleve',
    },
    ENVIRONMENT: {
        WEB_SERVER: 'environment.web_server',
        DATABASE: 'environment.database',
        ELASTICSEARCH: 'environment.elasticsearch',
        FILE_STORAGE: 'environment.file_storage',
        IMAGE_PROXY: 'environment.image_proxy',
        SMTP: 'environment.smtp',
        PUSH_NOTIFICATION_SERVER: 'environment.push_notification_server',
        HIGH_AVAILABILITY: 'environment.high_availability',
        RATE_LIMITING: 'environment.rate_limiting',
        LOGGING: 'environment.logging',
        SESSION_LENGTHS: 'environment.session_lengths',
        PERFORMANCE_MONITORING: 'environment.performance_monitoring',
        DEVELOPER: 'environment.developer',
        MOBILE_SECURITY: 'environment.mobile_security',
    },
};
exports.ResourceToSysConsolePermissionsTable = {
    [exports.RESOURCE_KEYS.ABOUT.EDITION_AND_LICENSE]: [permissions_1.default.SYSCONSOLE_READ_ABOUT_EDITION_AND_LICENSE, permissions_1.default.SYSCONSOLE_WRITE_ABOUT_EDITION_AND_LICENSE],
    billing: [permissions_1.default.SYSCONSOLE_READ_BILLING, permissions_1.default.SYSCONSOLE_WRITE_BILLING],
    [exports.RESOURCE_KEYS.REPORTING.SITE_STATISTICS]: [permissions_1.default.SYSCONSOLE_READ_REPORTING_SITE_STATISTICS, permissions_1.default.SYSCONSOLE_WRITE_REPORTING_SITE_STATISTICS],
    [exports.RESOURCE_KEYS.REPORTING.TEAM_STATISTICS]: [permissions_1.default.SYSCONSOLE_READ_REPORTING_TEAM_STATISTICS, permissions_1.default.SYSCONSOLE_WRITE_REPORTING_TEAM_STATISTICS],
    [exports.RESOURCE_KEYS.REPORTING.SERVER_LOGS]: [permissions_1.default.SYSCONSOLE_READ_REPORTING_SERVER_LOGS, permissions_1.default.SYSCONSOLE_WRITE_REPORTING_SERVER_LOGS],
    [exports.RESOURCE_KEYS.USER_MANAGEMENT.USERS]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_USERS, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_USERS],
    [exports.RESOURCE_KEYS.USER_MANAGEMENT.GROUPS]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_GROUPS, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_GROUPS],
    [exports.RESOURCE_KEYS.USER_MANAGEMENT.TEAMS]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_TEAMS, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_TEAMS],
    [exports.RESOURCE_KEYS.USER_MANAGEMENT.CHANNELS]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_CHANNELS, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_CHANNELS],
    [exports.RESOURCE_KEYS.USER_MANAGEMENT.PERMISSIONS]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_PERMISSIONS, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_PERMISSIONS],
    [exports.RESOURCE_KEYS.USER_MANAGEMENT.SYSTEM_ROLES]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_SYSTEM_ROLES, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_SYSTEM_ROLES],
    [exports.RESOURCE_KEYS.SYSTEM_ATTRIBUTES.USER_ATTRIBUTES]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_USERS, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_USERS],
    [exports.RESOURCE_KEYS.SYSTEM_ATTRIBUTES.ATTRIBUTE_BASED_ACCESS_CONTROL]: [permissions_1.default.SYSCONSOLE_READ_USERMANAGEMENT_SYSTEM_ROLES, permissions_1.default.SYSCONSOLE_WRITE_USERMANAGEMENT_SYSTEM_ROLES],
    [exports.RESOURCE_KEYS.SITE.CUSTOMIZATION]: [permissions_1.default.SYSCONSOLE_READ_SITE_CUSTOMIZATION, permissions_1.default.SYSCONSOLE_WRITE_SITE_CUSTOMIZATION],
    [exports.RESOURCE_KEYS.SITE.LOCALIZATION]: [permissions_1.default.SYSCONSOLE_READ_SITE_LOCALIZATION, permissions_1.default.SYSCONSOLE_WRITE_SITE_LOCALIZATION],
    [exports.RESOURCE_KEYS.SITE.USERS_AND_TEAMS]: [permissions_1.default.SYSCONSOLE_READ_SITE_USERS_AND_TEAMS, permissions_1.default.SYSCONSOLE_WRITE_SITE_USERS_AND_TEAMS],
    [exports.RESOURCE_KEYS.SITE.NOTIFICATIONS]: [permissions_1.default.SYSCONSOLE_READ_SITE_NOTIFICATIONS, permissions_1.default.SYSCONSOLE_WRITE_SITE_NOTIFICATIONS],
    [exports.RESOURCE_KEYS.SITE.ANNOUNCEMENT_BANNER]: [permissions_1.default.SYSCONSOLE_READ_SITE_ANNOUNCEMENT_BANNER, permissions_1.default.SYSCONSOLE_WRITE_SITE_ANNOUNCEMENT_BANNER],
    [exports.RESOURCE_KEYS.SITE.EMOJI]: [permissions_1.default.SYSCONSOLE_READ_SITE_EMOJI, permissions_1.default.SYSCONSOLE_WRITE_SITE_EMOJI],
    [exports.RESOURCE_KEYS.SITE.POSTS]: [permissions_1.default.SYSCONSOLE_READ_SITE_POSTS, permissions_1.default.SYSCONSOLE_WRITE_SITE_POSTS],
    [exports.RESOURCE_KEYS.SITE.FILE_SHARING_AND_DOWNLOADS]: [permissions_1.default.SYSCONSOLE_READ_SITE_FILE_SHARING_AND_DOWNLOADS, permissions_1.default.SYSCONSOLE_WRITE_SITE_FILE_SHARING_AND_DOWNLOADS],
    [exports.RESOURCE_KEYS.SITE.PUBLIC_LINKS]: [permissions_1.default.SYSCONSOLE_READ_SITE_PUBLIC_LINKS, permissions_1.default.SYSCONSOLE_WRITE_SITE_PUBLIC_LINKS],
    [exports.RESOURCE_KEYS.SITE.NOTICES]: [permissions_1.default.SYSCONSOLE_READ_SITE_NOTICES, permissions_1.default.SYSCONSOLE_WRITE_SITE_NOTICES],
    [exports.RESOURCE_KEYS.SITE.IP_FILTERING]: [permissions_1.default.SYSCONSOLE_READ_SITE_IP_FILTERS, permissions_1.default.SYSCONSOLE_WRITE_SITE_IP_FILTERS],
    [exports.RESOURCE_KEYS.ENVIRONMENT.WEB_SERVER]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_WEB_SERVER, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_WEB_SERVER],
    [exports.RESOURCE_KEYS.ENVIRONMENT.DATABASE]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_DATABASE, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_DATABASE],
    [exports.RESOURCE_KEYS.ENVIRONMENT.ELASTICSEARCH]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_ELASTICSEARCH, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_ELASTICSEARCH],
    [exports.RESOURCE_KEYS.ENVIRONMENT.FILE_STORAGE]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_FILE_STORAGE, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_FILE_STORAGE],
    [exports.RESOURCE_KEYS.ENVIRONMENT.IMAGE_PROXY]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_IMAGE_PROXY, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_IMAGE_PROXY],
    [exports.RESOURCE_KEYS.ENVIRONMENT.SMTP]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_SMTP, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_SMTP],
    [exports.RESOURCE_KEYS.ENVIRONMENT.PUSH_NOTIFICATION_SERVER]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_PUSH_NOTIFICATION_SERVER, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_PUSH_NOTIFICATION_SERVER],
    [exports.RESOURCE_KEYS.ENVIRONMENT.HIGH_AVAILABILITY]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_HIGH_AVAILABILITY, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_HIGH_AVAILABILITY],
    [exports.RESOURCE_KEYS.ENVIRONMENT.RATE_LIMITING]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_RATE_LIMITING, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_RATE_LIMITING],
    [exports.RESOURCE_KEYS.ENVIRONMENT.LOGGING]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_LOGGING, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_LOGGING],
    [exports.RESOURCE_KEYS.ENVIRONMENT.SESSION_LENGTHS]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_SESSION_LENGTHS, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_SESSION_LENGTHS],
    [exports.RESOURCE_KEYS.ENVIRONMENT.PERFORMANCE_MONITORING]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_PERFORMANCE_MONITORING, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_PERFORMANCE_MONITORING],
    [exports.RESOURCE_KEYS.ENVIRONMENT.DEVELOPER]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_DEVELOPER, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_DEVELOPER],
    [exports.RESOURCE_KEYS.ENVIRONMENT.MOBILE_SECURITY]: [permissions_1.default.SYSCONSOLE_READ_ENVIRONMENT_MOBILE_SECURITY, permissions_1.default.SYSCONSOLE_WRITE_ENVIRONMENT_MOBILE_SECURITY],
    [exports.RESOURCE_KEYS.AUTHENTICATION.SIGNUP]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_SIGNUP, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_SIGNUP],
    [exports.RESOURCE_KEYS.AUTHENTICATION.EMAIL]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_EMAIL, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_EMAIL],
    [exports.RESOURCE_KEYS.AUTHENTICATION.PASSWORD]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_PASSWORD, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_PASSWORD],
    [exports.RESOURCE_KEYS.AUTHENTICATION.MFA]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_MFA, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_MFA],
    [exports.RESOURCE_KEYS.AUTHENTICATION.LDAP]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_LDAP, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_LDAP],
    [exports.RESOURCE_KEYS.AUTHENTICATION.SAML]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_SAML, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_SAML],
    [exports.RESOURCE_KEYS.AUTHENTICATION.OPENID]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_OPENID, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_OPENID],
    [exports.RESOURCE_KEYS.AUTHENTICATION.GUEST_ACCESS]: [permissions_1.default.SYSCONSOLE_READ_AUTHENTICATION_GUEST_ACCESS, permissions_1.default.SYSCONSOLE_WRITE_AUTHENTICATION_GUEST_ACCESS],
    plugins: [permissions_1.default.SYSCONSOLE_READ_PLUGINS, permissions_1.default.SYSCONSOLE_WRITE_PLUGINS],
    [exports.RESOURCE_KEYS.INTEGRATIONS.INTEGRATION_MANAGEMENT]: [permissions_1.default.SYSCONSOLE_READ_INTEGRATIONS_INTEGRATION_MANAGEMENT, permissions_1.default.SYSCONSOLE_WRITE_INTEGRATIONS_INTEGRATION_MANAGEMENT],
    [exports.RESOURCE_KEYS.PRODUCTS.BOARDS]: [permissions_1.default.SYSCONSOLE_READ_PRODUCTS_BOARDS, permissions_1.default.SYSCONSOLE_WRITE_PRODUCTS_BOARDS],
    [exports.RESOURCE_KEYS.INTEGRATIONS.BOT_ACCOUNTS]: [permissions_1.default.SYSCONSOLE_READ_INTEGRATIONS_BOT_ACCOUNTS, permissions_1.default.SYSCONSOLE_WRITE_INTEGRATIONS_BOT_ACCOUNTS],
    [exports.RESOURCE_KEYS.INTEGRATIONS.GIF]: [permissions_1.default.SYSCONSOLE_READ_INTEGRATIONS_GIF, permissions_1.default.SYSCONSOLE_WRITE_INTEGRATIONS_GIF],
    [exports.RESOURCE_KEYS.INTEGRATIONS.CORS]: [permissions_1.default.SYSCONSOLE_READ_INTEGRATIONS_CORS, permissions_1.default.SYSCONSOLE_WRITE_INTEGRATIONS_CORS],
    [exports.RESOURCE_KEYS.COMPLIANCE.DATA_RETENTION_POLICY]: [permissions_1.default.SYSCONSOLE_READ_COMPLIANCE_DATA_RETENTION_POLICY, permissions_1.default.SYSCONSOLE_WRITE_COMPLIANCE_DATA_RETENTION_POLICY],
    [exports.RESOURCE_KEYS.COMPLIANCE.COMPLIANCE_EXPORT]: [permissions_1.default.SYSCONSOLE_READ_COMPLIANCE_COMPLIANCE_EXPORT, permissions_1.default.SYSCONSOLE_WRITE_COMPLIANCE_COMPLIANCE_EXPORT],
    [exports.RESOURCE_KEYS.COMPLIANCE.COMPLIANCE_MONITORING]: [permissions_1.default.SYSCONSOLE_READ_COMPLIANCE_COMPLIANCE_MONITORING, permissions_1.default.SYSCONSOLE_WRITE_COMPLIANCE_COMPLIANCE_MONITORING],
    [exports.RESOURCE_KEYS.COMPLIANCE.CUSTOM_TERMS_OF_SERVICE]: [permissions_1.default.SYSCONSOLE_READ_COMPLIANCE_CUSTOM_TERMS_OF_SERVICE, permissions_1.default.SYSCONSOLE_WRITE_COMPLIANCE_CUSTOM_TERMS_OF_SERVICE],
    [exports.RESOURCE_KEYS.EXPERIMENTAL.FEATURES]: [permissions_1.default.SYSCONSOLE_READ_EXPERIMENTAL_FEATURES, permissions_1.default.SYSCONSOLE_WRITE_EXPERIMENTAL_FEATURES],
    [exports.RESOURCE_KEYS.EXPERIMENTAL.FEATURE_FLAGS]: [permissions_1.default.SYSCONSOLE_READ_EXPERIMENTAL_FEATURE_FLAGS, permissions_1.default.SYSCONSOLE_WRITE_EXPERIMENTAL_FEATURE_FLAGS],
    [exports.RESOURCE_KEYS.EXPERIMENTAL.BLEVE]: [permissions_1.default.SYSCONSOLE_READ_EXPERIMENTAL_BLEVE, permissions_1.default.SYSCONSOLE_WRITE_EXPERIMENTAL_BLEVE],
};
