// mattermost-redux/lib/constants/telemetry.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventCategory = exports.eventSKUs = exports.TrackPropertyUserAgent = exports.TrackPropertyUser = exports.TrackMiscCategory = exports.TrackActionCategory = exports.TrackCrossTeamSearchAllTeamsEvent = exports.TrackCrossTeamSearchDifferentTeamEvent = exports.TrackCrossTeamSearchCurrentTeamEvent = exports.TrackPassiveKeywordsEvent = exports.TrackInviteGroupEvent = exports.TrackCrossTeamSearchFeature = exports.TrackScheduledPostsFeature = exports.TrackPassiveKeywordsFeature = exports.TrackGroupsFeature = exports.TrackEnterpriseSKU = exports.TrackProfessionalSKU = void 0;
// SKUs
exports.TrackProfessionalSKU = 'professional';
exports.TrackEnterpriseSKU = 'enterprise';
// Features
exports.TrackGroupsFeature = 'custom_groups';
exports.TrackPassiveKeywordsFeature = 'passive_keywords';
exports.TrackScheduledPostsFeature = 'scheduled_posts';
exports.TrackCrossTeamSearchFeature = 'cross_team_search';
// Events
exports.TrackInviteGroupEvent = 'invite_group_to_channel__add_member';
exports.TrackPassiveKeywordsEvent = 'update_passive_keywords';
exports.TrackCrossTeamSearchCurrentTeamEvent = 'cross_team_search__current_team';
exports.TrackCrossTeamSearchDifferentTeamEvent = 'cross_team_search__different_team';
exports.TrackCrossTeamSearchAllTeamsEvent = 'cross_team_search__all_teams';
// Categories
exports.TrackActionCategory = 'action';
exports.TrackMiscCategory = 'miscellaneous';
// Properties
exports.TrackPropertyUser = 'user_actual_id';
exports.TrackPropertyUserAgent = 'user_agent';
exports.eventSKUs = {
    [exports.TrackInviteGroupEvent]: [exports.TrackProfessionalSKU, exports.TrackEnterpriseSKU],
    [exports.TrackPassiveKeywordsEvent]: [exports.TrackProfessionalSKU, exports.TrackEnterpriseSKU],
};
exports.eventCategory = {
    [exports.TrackInviteGroupEvent]: exports.TrackActionCategory,
    [exports.TrackPassiveKeywordsEvent]: exports.TrackActionCategory,
    [exports.TrackCrossTeamSearchAllTeamsEvent]: exports.TrackActionCategory,
    [exports.TrackCrossTeamSearchCurrentTeamEvent]: exports.TrackActionCategory,
    [exports.TrackCrossTeamSearchDifferentTeamEvent]: exports.TrackActionCategory,
};
