// mattermost-redux/lib/reducers/requests/index.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const redux_1 = require("redux");
const admin_1 = __importDefault(require("./admin"));
const channels_1 = __importDefault(require("./channels"));
const files_1 = __importDefault(require("./files"));
const general_1 = __importDefault(require("./general"));
const posts_1 = __importDefault(require("./posts"));
const roles_1 = __importDefault(require("./roles"));
const search_1 = __importDefault(require("./search"));
const teams_1 = __importDefault(require("./teams"));
const users_1 = __importDefault(require("./users"));
exports.default = (0, redux_1.combineReducers)({
    channels: channels_1.default,
    files: files_1.default,
    general: general_1.default,
    posts: posts_1.default,
    teams: teams_1.default,
    users: users_1.default,
    admin: admin_1.default,
    search: search_1.default,
    roles: roles_1.default,
});
