// mattermost-redux/lib/actions/index.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.users = exports.websocket = exports.timezone = exports.teams = exports.search = exports.schemes = exports.roles = exports.preferences = exports.posts = exports.jobs = exports.helpers = exports.integrations = exports.groups = exports.general = exports.files = exports.emojis = exports.errors = exports.cloud = exports.channels = exports.bots = exports.admin = void 0;
const admin = __importStar(require("./admin"));
exports.admin = admin;
const bots = __importStar(require("./bots"));
exports.bots = bots;
const channels = __importStar(require("./channels"));
exports.channels = channels;
const cloud = __importStar(require("./cloud"));
exports.cloud = cloud;
const emojis = __importStar(require("./emojis"));
exports.emojis = emojis;
const errors = __importStar(require("./errors"));
exports.errors = errors;
const files = __importStar(require("./files"));
exports.files = files;
const general = __importStar(require("./general"));
exports.general = general;
const groups = __importStar(require("./groups"));
exports.groups = groups;
const helpers = __importStar(require("./helpers"));
exports.helpers = helpers;
const integrations = __importStar(require("./integrations"));
exports.integrations = integrations;
const jobs = __importStar(require("./jobs"));
exports.jobs = jobs;
const posts = __importStar(require("./posts"));
exports.posts = posts;
const preferences = __importStar(require("./preferences"));
exports.preferences = preferences;
const roles = __importStar(require("./roles"));
exports.roles = roles;
const schemes = __importStar(require("./schemes"));
exports.schemes = schemes;
const search = __importStar(require("./search"));
exports.search = search;
const teams = __importStar(require("./teams"));
exports.teams = teams;
const timezone = __importStar(require("./timezone"));
exports.timezone = timezone;
const users = __importStar(require("./users"));
exports.users = users;
const websocket = __importStar(require("./websocket"));
exports.websocket = websocket;
